integration/ingester_test.go

```diff
@@ -0,0 +1,336 @@
+						},
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep * 5).UnixMilli()),
+							Histogram: GenerateTestSampleHistogram(6),
+						},
+					},
+				},
+			},
+		},
+		"series switching from float to native histograms at timestamp": {
+			inSeries: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{
+							Name:  "__name__",
+							Value: query,
+						},
+					},
+					Samples: []prompb.Sample{
+						{
+							Timestamp: queryStart.UnixMilli(),
+							Value:     100,
+						},
+						{
+							Timestamp: queryStart.Add(queryStep).UnixMilli(),
+							Value:     110,
+						},
+						{
+							Timestamp: queryStart.Add(queryStep * 2).UnixMilli(),
+							Value:     120,
+						},
+					},
+					Histograms: []prompb.Histogram{
+						remote.HistogramToHistogramProto(queryStart.Add(queryStep*3).UnixMilli(), tsdb.GenerateTestHistogram(4)),
+						remote.HistogramToHistogramProto(queryStart.Add(queryStep*4).UnixMilli(), tsdb.GenerateTestHistogram(5)),
+						remote.HistogramToHistogramProto(queryStart.Add(queryStep*5).UnixMilli(), tsdb.GenerateTestHistogram(6)),
+					},
+				},
+			},
+			expected: model.Matrix{
+				{
+					Metric: model.Metric(map[model.LabelName]model.LabelValue{"__name__": "foobar"}),
+					Values: []model.SamplePair{
+						{
+							Timestamp: model.Time(queryStart.UnixMilli()),
+							Value:     model.SampleValue(100),
+						},
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep).UnixMilli()),
+							Value:     model.SampleValue(110),
+						},
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep * 2).UnixMilli()),
+							Value:     model.SampleValue(120),
+						},
+					},
+					Histograms: []model.SampleHistogramPair{
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep * 3).UnixMilli()),
+							Histogram: GenerateTestSampleHistogram(4),
+						},
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep * 4).UnixMilli()),
+							Histogram: GenerateTestSampleHistogram(5),
+						},
+						{
+							Timestamp: model.Time(queryStart.Add(queryStep * 5).UnixMilli()),
+							Histogram: GenerateTestSampleHistogram(6),
+						},
+					},
+				},
+			},
+		},
+		"series including float and native histograms at same timestamp": {
+			inSeries: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{
+							Name:  "__name__",
+							Value: query,
+						},
+					},
+					Samples: []prompb.Sample{
+						{
+							Timestamp: queryStart.UnixMilli(),
+							Value:     100,
+						},
+						{
+							Timestamp: queryStart.Add(queryStep).UnixMilli(),
+							Value:     110,
+						},
+						{
+							// This will get overwritten by the histogram with the same timestamp.
+							Timestamp: queryStart.Add(queryStep * 2).UnixMilli(),
+							Value:     120,
+						},
+					},
+					Histograms: []prompb.Histogram{
+						remote.HistogramToHistogramProto(queryStart.Add(queryStep*2).UnixMilli(), tsdb.GenerateTestHistogram(3)),
```

Comment:
I am curious. Is this sample not erroring out because the timestamp matches the last sample? If the time was let's say even 1ms older, will this turn into an out-of-order sample? This looks like a bug upstream, this should not be ingestible without out-of-order enabled.

Reply:
Oops, yeah you are correct and the comment in this test was wrong; I'm updating the comment to correctly reflect which sample fails to get appended.